src/sync/app.cpp

```
@@ -710,23 +710,69 @@ std::string App::url_for_path(const std::string& path="") const
+        .url = route,
+        .timeout_ms = m_request_timeout_ms
+    };
+
+    m_config.transport_generator()->send_request_to_server(req, [this, completion_block](const Response& response) {
+        nlohmann::json json;
+        try {
+            json = nlohmann::json::parse(response.body);
+        } catch (const std::exception& e) {
+            return completion_block(AppError(make_error_code(JSONErrorCode::malformed_json), e.what()));
+        }
+
+        try {
+            this->m_metadata = AppMetadata(
+                value_from_json<std::string>(json, "deployment_model"),
+                value_from_json<std::string>(json, "location"),
+                value_from_json<std::string>(json, "hostname"),
+                value_from_json<std::string>(json, "ws_hostname"));
+            m_base_route = m_metadata->m_hostname + default_base_path;
+            m_app_route = (m_base_route + app_path + "/" + m_config.app_id);
+            m_auth_route = (m_app_route + auth_path);
+        } catch (const AppError& err) {
+            return completion_block(err);
+        }
+
+        completion_block(util::none);
+    });
+}
+
 void App::do_authenticated_request(Request request,
                                    std::shared_ptr<SyncUser> sync_user,
-                                   std::function<void (Response)> completion_block) const
+                                   std::function<void (Response)> completion_block)
 {
-    auto handler = [completion_block, request, sync_user, this](const Response& response) {
-        if (auto error = check_for_errors(response)) {
-            App::handle_auth_failure(error.value(), response, request, sync_user, completion_block);
-        } else {
-            completion_block(response);
+    init_app_metadata([completion_block, &request, sync_user, this](const util::Optional<AppError> error){
+        if (error) {
+            return completion_block(Response{.http_status_code = 400});
```

Shouldn't we return the real error here, instead of hiding it?